kernel/ipi.c

```diff
@@ -13,15 +13,62 @@ extern void z_trace_sched_ipi(void);
 #endif
 
 
-void flag_ipi(void)
+void flag_ipi(uint32_t ipi_mask)
 {
 #if defined(CONFIG_SCHED_IPI_SUPPORTED)
 	if (arch_num_cpus() > 1) {
-		_kernel.pending_ipi = true;
+		atomic_or(&_kernel.pending_ipi, (atomic_val_t)ipi_mask);
 	}
 #endif /* CONFIG_SCHED_IPI_SUPPORTED */
 }
 
+/* Create a bitmask of CPUs that need an IPI. Note: sched_spinlock is held. */
+atomic_val_t ipi_mask_create(struct k_thread *thread)
+{
+#if defined(CONFIG_IPI_OPTIMIZE)
+	uint32_t  ipi_mask = 0;
+	uint32_t  num_cpus = (uint32_t)arch_num_cpus();
+	uint32_t  id = _current_cpu->id;
+	struct k_thread *cpu_thread;
+
+	for (uint32_t i = 0; i < num_cpus; i++) {
+		if (id == i) {
+			continue;
+		}
+
+		/*
+		 * An IPI absolutely does not need to be sent if ...
+		 * 1. the CPU is not active, or
+		 * 2. <thread> can not execute on the target CPU
+		 * ... and might not need to be sent if ...
+		 * 3. the target CPU's active thread is not preemptible, or
+		 * 4. the target CPU's active thread has a higher priority
+		 *    (Items 3 & 4 may be overridden by a metaIRQ thread)
+		 */
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
+#if defined(CONFIG_SCHED_CPU_MASK)
+		    ((thread->base.cpu_mask & BIT(i)) != 0) &&
+#endif
+		    (thread_is_metairq(thread) ||
+		     (thread_is_preemptible(cpu_thread) &&
+		      (z_sched_prio_cmp(cpu_thread, thread) < 0)))) {
```

Comment:
Style nit:  please no #if around subexpresssions.  Ew :).  Abstract the cpu_mask test into a predicate function like the other steps here and put the preprocessor logic there.

Performance nit: I'd reorder this.   First to remove the test for NULL, which should never happen in a live kernel[1].  After that IMHO by far the most common decision case is the simple priority comparison, then followed by preemptible priorities with metairq last (which should be rare in almost all circumstances, given that they block the world apps will work hard to make them quick).  I'd put cpu_mask last too, but there's some argument as whether or not it's a common case is app-dependent; but my argument would be that apps that use cpu_mask a lot (like SOF) wouldn't use this feature anyway as they will be doing coarse scheduling at the app level (e.g. SOF basically has only one thread per core and doesn't do much kernel scheduling).

[1]  Under no circumstances should a CPU receive an IPI before it's been initialized (i.e. everything should be set before it unmasks interrupts for the first time).  Maybe replace with an assert if you're worried about a specific edge case?

Reply:
I am happy to reorder the checks. However, I do think that the test for NULL is necessary. It ensures that we do not do not dereference a NULL pointer. Recall that static threads get initialized (and may start) before the other CPUs are initialized. Uninitialized CPUs have their _current_ field set to NULL. When these threads start, they can flag IPIs which can then get signalled.

Reply:
Ah, indeed.  This is the sending case, not the receiving case, and not all the cpus may be online at a given time (intel_adsp can actually shut them down). 

Which actually brings up a good question: do we null-out the current field of an offline CPU anywhere?  I think we might not, in which case this is going to be stuck throwing IPIs at dead CPUs (which is... probably benign in most cases?  I guess it depends on architecture).  We don't really have a single authoritative place to answer\ "is this CPU alive?".  Not something for this PR but almost certainly something on which we'll stumble in the future.